autoload/ale/autocmd.vim

```diff
@@ -1,77 +1,62 @@
-    let l:text_changed = '' . g:ale_lint_on_text_changed
+    if g:ale_enabled
+        " This value used to be a Boolean as a Number, and is now a String.
+        let l:text_changed = '' . g:ale_lint_on_text_changed
 
-    augroup ALEPatternOptionsGroup
-        autocmd!
-        autocmd BufEnter,BufRead * call ale#pattern_options#SetOptions(str2nr(expand('<abuf>')))
-    augroup END
-
-    augroup ALERunOnTextChangedGroup
-        autocmd!
-        if g:ale_enabled
+        augroup ALETriggerGroup
             if l:text_changed is? 'always' || l:text_changed is# '1'
                 autocmd TextChanged,TextChangedI * call ale#Queue(g:ale_lint_delay)
             elseif l:text_changed is? 'normal'
                 autocmd TextChanged * call ale#Queue(g:ale_lint_delay)
             elseif l:text_changed is? 'insert'
                 autocmd TextChangedI * call ale#Queue(g:ale_lint_delay)
             endif
-        endif
-    augroup END
 
-    augroup ALERunOnEnterGroup
-        autocmd!
-        if g:ale_enabled
+            if g:ale_lint_on_insert_leave
+                autocmd InsertLeave * call ale#Queue(0)
+            endif
+
             " Handle everything that needs to happen when buffers are entered.
             autocmd BufEnter * call ale#events#EnterEvent(str2nr(expand('<abuf>')))
-        endif
-        if g:ale_enabled && g:ale_lint_on_enter
-            autocmd BufWinEnter,BufRead * call ale#Queue(0, 'lint_file', str2nr(expand('<abuf>')))
-            " Track when the file is changed outside of Vim.
-            autocmd FileChangedShellPost * call ale#events#FileChangedEvent(str2nr(expand('<abuf>')))
-        endif
-    augroup END
-
-    augroup ALERunOnFiletypeChangeGroup
-        autocmd!
-        if g:ale_enabled && g:ale_lint_on_filetype_changed
-            " Only start linting if the FileType actually changes after
-            " opening a buffer. The FileType will fire when buffers are opened.
-            autocmd FileType * call ale#events#FileTypeEvent(
-            \   str2nr(expand('<abuf>')),
-            \   expand('<amatch>')
-            \)
-        endif
-    augroup END
-
-    augroup ALERunOnSaveGroup
-        autocmd!
-        autocmd BufWritePost * call ale#events#SaveEvent(str2nr(expand('<abuf>')))
-    augroup END
+            if g:ale_lint_on_enter
+                autocmd BufWinEnter,BufRead * call ale#Queue(0, 'lint_file', str2nr(expand('<abuf>')))
+                " Track when the file is changed outside of Vim.
+                autocmd FileChangedShellPost * call ale#events#FileChangedEvent(str2nr(expand('<abuf>')))
+            endif
 
-    augroup ALERunOnInsertLeave
-        autocmd!
-        if g:ale_enabled && g:ale_lint_on_insert_leave
-            autocmd InsertLeave * call ale#Queue(0)
-        endif
-    augroup END
+            if g:ale_lint_on_save
+                autocmd BufWritePost * call ale#events#SaveEvent(str2nr(expand('<abuf>')))
```

Comment:
This event will still need to be there if ALE fixes something on save, when `g:ale_enabled` is `0`. The `fix_on_save` setting could only be on for a particular buffer, so we probably have to always have this event set up.